src/search/handler.h

```diff
@@ -0,0 +1,67 @@
+﻿/*
+===========================================================================
+
+  Copyright (c) 2024 LandSandBoat Dev Teams
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see http://www.gnu.org/licenses/
+
+===========================================================================
+*/
+
+#pragma once
+
+#include <asio/ts/buffer.hpp>
+#include <asio/ts/internet.hpp>
+
+#include "search_session.h"
+
+template <typename T>
+class handler
+{
+public:
+    handler(asio::io_context& io_context, unsigned int port)
+    : acceptor_(io_context, asio::ip::tcp::endpoint(asio::ip::tcp::v4(), port))
+    {
+        acceptor_.set_option(asio::socket_base::reuse_address(true));
+
+        do_accept();
+    }
+
+private:
+    void do_accept()
+    {
+        // clang-format off
+        acceptor_.async_accept(
+        [this](std::error_code ec, asio::ip::tcp::socket socket)
+        {
+            if (!ec)
+            {
+                if constexpr (std::is_same_v<T, search_session>)
+                {
+                    auto auth_handler = std::make_shared<T>(std::move(socket));
```

Comment:
Should this be search_handler? This is the thing that's doing shared_from_this() all the way down the chain right?

Reply:
done